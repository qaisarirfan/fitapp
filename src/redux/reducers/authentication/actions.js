import {createActionName} from "../../utility"

export const reducerName = "authentication"

// Actions
export const LOGIN = createActionName(reducerName, "LOGIN")
export const LOGOUT = createActionName(reducerName, "LOGOUT")
