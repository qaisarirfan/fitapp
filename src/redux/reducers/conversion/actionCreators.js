import {
  CHANGE_BASE_CURRENCY,
  CHANGE_QUOTE_CURRENCY,
  RATES,
  SWAP_CURRENCY,
  TOGGLE_FAVORITE,
} from "./actions"

// conversion actions
export const createLoadAction = (base = "USD") => ({
  type: RATES,
  request: {
    type: "get",
    url: `/latest?base=${base}`,
  },
})

export const swapCurrency = () => ({
  type: SWAP_CURRENCY,
})

export const changeBaseCurrency = (currency) => ({
  type: CHANGE_BASE_CURRENCY,
  payload: currency,
})

export const changeQuoteCurrency = (currency) => ({
  type: CHANGE_QUOTE_CURRENCY,
  payload: currency,
})

export const toggleFavorite = (currency) => ({
  type: TOGGLE_FAVORITE,
  payload: currency,
})
