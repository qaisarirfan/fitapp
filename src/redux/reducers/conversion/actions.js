import {createActionName} from "../../utility"

// conversion name
export const reducerName = "conversion"

// conversion actions type
export const RATES = createActionName(reducerName, "RATES")
export const SWAP_CURRENCY = "SWAP_CURRENCY"
export const CHANGE_BASE_CURRENCY = "CHANGE_BASE_CURRENCY"
export const CHANGE_QUOTE_CURRENCY = "CHANGE_QUOTE_CURRENCY"
export const TOGGLE_FAVORITE = "TOGGLE_FAVORITE"
