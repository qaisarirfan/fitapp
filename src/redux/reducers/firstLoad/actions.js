import {createActionName} from "../../utility"

export const reducerName = "firstLoad"

// Actions
export const LOAD = createActionName(reducerName, "LOAD")
export const SHOW_TUTORIAL = createActionName(reducerName, "SHOW_TUTORIAL")
export const HIDE_TUTORIAL = createActionName(reducerName, "HIDE_TUTORIAL")
