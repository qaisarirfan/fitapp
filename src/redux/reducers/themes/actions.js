import {createActionName} from "../../utility"

export const reducerName = "themes"

// Actions
export const SELECT_THEME = createActionName(reducerName, "SELECT_THEME")
