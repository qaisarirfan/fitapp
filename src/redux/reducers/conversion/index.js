import get from "lodash/get"
import findIndex from "lodash/findIndex"
import {createReducer} from "../../utility"
import {
  CHANGE_BASE_CURRENCY,
  CHANGE_QUOTE_CURRENCY,
  RATES,
  SWAP_CURRENCY,
  TOGGLE_FAVORITE,
} from "./actions"
import {ERROR, LOADED, LOADING} from "../../middleware/actions"
import {LOGOUT} from "../authentication/actions"

// conversion initial state
export const initialState = {
  baseCurrency: "USD",
  quoteCurrency: "GBP",
  rates: {
    data: [],
    loader: false,
    loadingError: null,
  },
  favorite: [],
}

// conversion Reducer
const reducers = {
  [RATES + LOADING](state) {
    return {
      ...state,
      rates: {
        ...state.rates,
        data: [],
        loadingError: null,
        loader: true,
      },
    }
  },

  [RATES + LOADED](state, payload) {
    const result = get(payload, "result.rates", {})
    return {
      ...state,
      rates: {
        ...state.rates,
        data: Object.keys(result).map((key) => ({
          name: key,
          rate: result[key],
          isSelected: false,
          isFavorite: false,
        })),
        loader: false,
      },
    }
  },

  [RATES + ERROR](state, payload) {
    const result = get(payload, "result", undefined)
    return {
      ...state,
      rates: {
        ...state.rates,
        data: [],
        loader: false,
        loadingError: result,
      },
    }
  },

  [SWAP_CURRENCY](state) {
    return {
      ...state,
      baseCurrency: state.quoteCurrency,
      quoteCurrency: state.baseCurrency,
    }
  },

  [CHANGE_BASE_CURRENCY](state, payload) {
    return {
      ...state,
      baseCurrency: payload,
    }
  },

  [CHANGE_QUOTE_CURRENCY](state, payload) {
    return {
      ...state,
      quoteCurrency: payload,
    }
  },

  [TOGGLE_FAVORITE](state, payload) {
    const favorite = [...state.favorite]

    const index = findIndex(state.favorite, (fav) => {
      return fav === payload
    })

    if (index === -1) {
      favorite.push(payload)
    } else {
      favorite.splice(index, 1)
    }

    return {
      ...state,
      favorite,
    }
  },

  [LOGOUT]() {
    return initialState
  },
}

export default createReducer(reducers, initialState)
