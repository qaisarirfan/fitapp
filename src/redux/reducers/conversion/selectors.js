import get from "lodash/get"
import findIndex from "lodash/findIndex"
import {createSelector} from "reselect"
import {reducerName} from "./actions"

// conversion selectors
export const selectRatesData = (state) =>
  get(state, `${reducerName}.rates.data`)
export const selectRatesLoader = (state) =>
  get(state, `${reducerName}.rates.loader`)
export const selectRatesLoadingError = (state) =>
  get(state, `${reducerName}.rates.loadingError`)

export const selectBaseCurrency = (state) =>
  get(state, `${reducerName}.baseCurrency`)
export const selectQuoteCurrency = (state) =>
  get(state, `${reducerName}.quoteCurrency`)
export const selectFavoriteCurrencies = (state) =>
  get(state, `${reducerName}.favorite`)

export const selectCurrencies = createSelector(
  selectRatesData,
  selectFavoriteCurrencies,
  (rates, favorite) => {
    return (
      rates &&
      rates.map((rate) => {
        const index = findIndex(favorite, (fav) => {
          return fav === rate.name
        })
        return {
          ...rate,
          isFavorite: index !== -1,
        }
      })
    )
  }
)
