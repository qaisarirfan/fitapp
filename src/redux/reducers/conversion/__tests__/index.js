import conversion, {initialState} from ".."
import { ERROR, LOADED, LOADING } from "../../../middleware/actions"
import { RATES } from "../actions"

describe("conversion reducer", () => {
  it("should return the initial state", () => {
    const result = expect(conversion(undefined, {}))
    result.toEqual(initialState)
    result.toMatchSnapshot()
  })
  it("should handle GET_REQUEST_START", () => {
    const result = expect(
      conversion(undefined, { 
        type: RATES+LOADING,
        request:{
          type:"get",
          url:"/latest"
        }
      })
    )
    result.toEqual({
      ...initialState,
      rates:{
        ...initialState.rates,
        loader:true
      }
    })
    result.toMatchSnapshot()
  })
  it("should handle GET_REQUEST_SUCCESS", () => {
    const result = expect(
      conversion(undefined, {
        type: RATES+LOADED,
        request:{
          type:"get",
          url:"/latest"
        }
      })
    )
    result.toEqual({
      ...initialState,
      rates:{
        ...initialState.rates,
        loader:false,
        data:[]
      }
    })
    result.toMatchSnapshot()
  })
  it("should handle GET_REQUEST_FAIL", () => {
    const result = expect(
      conversion(undefined, {
        type: RATES+ERROR,
        request:{
          type:"get",
          url:"/latest/asdf"
        }
      })
    )
    result.toEqual({
      ...initialState,
      rates:{
        ...initialState.rates,
        loader:false,
        loadingError:undefined
      }
    })
    result.toMatchSnapshot()
  })
})
