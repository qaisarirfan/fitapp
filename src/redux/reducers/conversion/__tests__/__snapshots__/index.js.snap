// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`conversion reducer should handle GET_REQUEST_FAIL 1`] = `
Object {
  "baseCurrency": "USD",
  "favorite": Array [],
  "quoteCurrency": "GBP",
  "rates": Object {
    "data": Array [],
    "loader": false,
    "loadingError": undefined,
  },
}
`;

exports[`conversion reducer should handle GET_REQUEST_START 1`] = `
Object {
  "baseCurrency": "USD",
  "favorite": Array [],
  "quoteCurrency": "GBP",
  "rates": Object {
    "data": Array [],
    "loader": true,
    "loadingError": null,
  },
}
`;

exports[`conversion reducer should handle GET_REQUEST_SUCCESS 1`] = `
Object {
  "baseCurrency": "USD",
  "favorite": Array [],
  "quoteCurrency": "GBP",
  "rates": Object {
    "data": Array [],
    "loader": false,
    "loadingError": null,
  },
}
`;

exports[`conversion reducer should return the initial state 1`] = `
Object {
  "baseCurrency": "USD",
  "favorite": Array [],
  "quoteCurrency": "GBP",
  "rates": Object {
    "data": Array [],
    "loader": false,
    "loadingError": null,
  },
}
`;
